Create EntityNotDeclared code action

xorye opened the pull request:
Fixes #532 

This PR adds a new code action to locally declare an undeclared entity.

```xml
<?xml version="1.0" encoding="UTF-8"?>
<article>
   &nbsp;
</article>
```

Demo:
![demo](https://raw.githubusercontent.com/xorye/gifs/master/pr/entity_not_declared_ca_2.gif?token=AE3CR5J2JHMDBMFLSFLICM2622T52)

The codeaction works whether or not `<!DOCTYPE` is declared or not, and whether the internal subset `[]` exists or not.

Signed-off-by: David Kwon <dakwon@redhat.com>

xorye commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java:
```diff
@@ -0,0 +1,185 @@
+
+	private String getEntityName(Diagnostic diagnostic, Range range, DOMDocument doc) throws BadLocationException {
+		String name = doc.getText().substring(doc.offsetAt(range.getStart()), doc.offsetAt(range.getEnd()));
+		String removedAmpAndSemiColon = name.substring(1, name.length() - 1);
```
Here, I am retrieving the entity name from the error range, and then checking it against the error message.
It would be better if we could store the name in the `Diagnostic` when the `Diagnostic` is being created.

Related issue: https://github.com/eclipse/lemminx/issues/681

angelozerr replied:
Indeed it's little ugly but for the moment we have not teh choice to do that. Please add a comments for that and set the link to the LSP issue (and please do that for the code #681)

xorye commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java:
```diff
@@ -0,0 +1,185 @@
+		insertString.append("<!DOCTYPE article [" + delimiter);
+		insertString.append(createEntityDeclaration(entityName, settings));
+		insertString.append(delimiter);
+		insertString.append("]>" + delimiter);
```
TBH is it okay to use string concatenation instead of StringBuilder in this PR? I think StringBuilder might not be necessary unless I am appending string content in a loop.

https://stackoverflow.com/questions/4645020/when-to-use-stringbuilder-in-java
https://stackoverflow.com/questions/5234147/why-stringbuilder-when-there-is-string
https://stackoverflow.com/questions/24571463/when-should-you-explicitly-use-a-stringbuilder

angelozerr replied:
To generate XML, it should be better to use XMLBuilder instead of StringBuilder.

datho7561 commented on org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/contentmodel/DTDDiagnosticsTest.java:
```diff
@@ -280,6 +285,143 @@ public void EntityDeclUnterminated() throws Exception {
+		XMLAssert.testDiagnosticsFor(xml, d);
+		testCodeActionsFor(xml, d, settings, ca(d, te(2, 29, 2, 29, "\r\n\t<!ENTITY nbsp \'entity-value\'>")));
+	}
+
```
```xml
<?xml version="1.0" encoding="UTF-8"?><text>
<text>
  &c;
</text>
</text>
```
The code action will put the DOCTYPE declaration after the first `<text>` tag, on the second line:

![make-DTD-entity-code-action](https://user-images.githubusercontent.com/22376627/83052570-82f7e000-a01d-11ea-8661-d18ad826f852.gif)

If it makes sense to fix this issue, then we should test for it here.

xorye replied:
Thanks, fixed

xorye commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java:
```diff
@@ -0,0 +1,185 @@
+		String delimiter = document.lineDelimiter(insertPosition.getLine());
+		String message = "Declare DOCTYPE containing ENTITY " + entityName;
+		StringBuilder insertString = new StringBuilder();
+		insertString.append("<!DOCTYPE article [" + delimiter);
```
Thanks, fixed. I think its a much better idea to get it from the top level element like you said

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/DOMDocument.java:
```diff
@@ -307,6 +307,24 @@ public boolean hasProlog() {
+	 * @return true if this document has a DTD DOCTYPE
+	 * declaration. Returns false otherwise.
+	 */
+	public boolean hasDoctype() {
```
I think you don"t need this method. Please remove it and use DOMDocumentType docType = document.getDoctype();.

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java:
```diff
@@ -0,0 +1,205 @@
+			if (entityName == null) {
+				return;
+			}
+			if (document.hasDoctype()) {
```
```
DOMDocumentType docType = document.getDoctype();
if (docType != null) {
```

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/DOMDocumentType.java:
```diff
@@ -212,7 +211,7 @@ void setSystemId(int start, int end) {
 	 */
 	public String getSubstring(int start, int end) {
-		return getContent().substring(start - getStart(), end - getStart());
+		return getTextContent().substring(start - getStart(), end - getStart());
```
It seems that getTextContent  could return null.

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java:
```diff
@@ -0,0 +1,205 @@
+				// Add <!ENTITY nbsp "entity-value"> in the subset.
+				// If the subset does not exist, add subset too
+				
+				// ie:
```
I think you have already spoken about the following usecase, but what about DOCTYPE without subset?

fbricon commented:
please rebase against master

xorye commented:
Thank you, fixed

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java:
```diff
@@ -0,0 +1,211 @@
+				addDoctypeAndEntityCodeAction(entityName, diagnostic, document, sharedSettings, codeActions);
+			}
+		} catch (BadLocationException e) {
+
```
Add LOGGER

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java:
```diff
@@ -0,0 +1,211 @@
+		}
+	}
+
+	private void addEntityCodeAction(String entityName, Diagnostic diagnostic, DOMDocument document,
```
add Javadoc

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java:
```diff
@@ -0,0 +1,211 @@
+
+		XMLBuilder insertString = new XMLBuilder(settings, null, delimiter);
+
+		if (docType.getInternalSubset() == null) {
```
```java
boolean hasInternalSubset = docType.getInternalSubset() != null;
if (!hasInternalSubset ) {
```

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java:
```diff
@@ -0,0 +1,211 @@
+		insertString.indent(1);
+		addEntityDeclaration(entityName, insertString);
+
+		if (docType.getInternalSubset() == null) {
```
`if (!hasInternalSubset ) {`

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java:
```diff
@@ -0,0 +1,211 @@
+		codeActions.add(action);
+	}
+
+	private void addDoctypeAndEntityCodeAction(String entityName, Diagnostic diagnostic, DOMDocument document,
```
Add Javadoc

xorye commented:
Fixed, I also added a second commit that adds tests for https://github.com/eclipse/lemminx/issues/686, since that issue is being fixed thanks to this new method:
https://github.com/eclipse/lemminx/blob/765bc8fe1959f2f54bea9667e6c4b8e6320ea450/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/utils/XMLBuilder.java#L423-L432

angelozerr commented on org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/pi-snippets.json:
```diff
@@ -0,0 +1,20 @@
+{
```
I think this file should not belong to this PR

xorye replied:
thanks, removed

angelozerr commented on org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/utils/StringUtils.java:
```diff
@@ -439,4 +439,47 @@ public static int findEndWord(String text, int offset, Predicate<Character> isVa
+			return false;
+		}
+		char quoteValueStart = value.charAt(0);
+		boolean start = quoteValueStart == '\"' || quoteValueStart == '\'' ? true : false;
+		if (start == false) {
+			return false;
+		}
```
+1

xorye commented:
Great feedback, fixed

angelozerr commented:
@xorye you have 2 commits for this PR. I think the commit https://github.com/eclipse/lemminx/pull/724/commits/084c611c35645c459a144980c03acaec2af013f3 should be removed, no?

fbricon commented:
@angelozerr see https://github.com/eclipse/lemminx/pull/724#issuecomment-636011370

angelozerr commented:
> @angelozerr see #724 (comment)

Thanks @fbricon

angelozerr commented:
Great PR @xorye !
